### PR TITLE
[Bugfix]Fix Hunyuan Image3 denoise flow alignment

### PR DESCRIPTION
**Pull request by bjf-frz**
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
  Align the Hunyuan Image3 DiT denoising flow with the native implementation to reduce precision divergence during multi-step image generation.

  This change fixes three flow-level mismatches:

  - Normalize system prompt whitespace before prompt/token construction.
  - Stop carrying `input_ids` after the first denoising forward, matching the native flow where later steps are driven by timestep/image embeddings, position IDs, attention
  masks, and KV cache.
  - Reuse image attention KV cache by scattering current-step K/V into the full first-step cache using absolute `position_ids`, instead of rebuilding K/V with an inferred
  prompt/image offset.

  For DiT-only precision validation, the Hunyuan Image3 DiT deploy config is also pinned to the Torch SDPA diffusion attention backend.

## Test Plan

accuracy test:
pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py

offline command:
python examples/offline_inference/hunyuan_image3/end2end.py \
    --model tencent/HunyuanImage-3.0-Instruct \
    --modality text2img \
    --deploy-config vllm_omni/deploy/hunyuan_image3_dit.yaml \
    --prompts "A brown and white dog is running on the grass." \
    --output ./output \
    --steps 50 \
    --guidance-scale 2.5 \
    --seed 42 \
    --height 1024 \
    --width 1024 \
    --bot-task none \
    --sys-type en_unified \
    --enforce-eager

## Test Result

Accuracy test:
before:
```
+-----------------------------+---------+------------------+
| Metric                      |   Value |   L20x Reference |
+=============================+=========+==================+
| COT similarity to reference |  0.9731 |           0.9644 |
+-----------------------------+---------+------------------+
| COT prefix match            | 29      |          29      |
+-----------------------------+---------+------------------+
| Image-Image similarity      | 92.8115 |          94.5538 |
+-----------------------------+---------+------------------+
| SSIM                        |  0.2549 |           0.242  |
+-----------------------------+---------+------------------+
| PSNR (dB)                   | 14      |          14.1    |
+-----------------------------+---------+------------------+
```

after:
```
+-----------------------------+---------+------------------+
| Metric                      |   Value |   L20x Reference |
+=============================+=========+==================+
| COT similarity to reference |  0.9751 |           0.9644 |
+-----------------------------+---------+------------------+
| COT prefix match            | 45      |          29      |
+-----------------------------+---------+------------------+
| Image-Image similarity      | 97.2919 |          94.5538 |
+-----------------------------+---------+------------------+
| SSIM                        |  0.2838 |           0.242  |
+-----------------------------+---------+------------------+
| PSNR (dB)                   | 14.16   |          14.1    |
+-----------------------------+---------+------------------+
PASSED
```

native output:
native_50

vllm-omni aligned output:
output_0_0_pr

vllm-omni aligned tp2 sp2 output:
output_0_0_tp2sp2

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [ ] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 1]
```diff
@@ -954,15 +958,27 @@ def _reuse_prompt_kv(
...
         """
         Reuse cached prompt KV in subsequent denoising steps.
 
-        Non-SP: returns [cached_prompt, new_image, zero_eoi] shaped [bs, seq_len, ...].
+        Non-SP: scatters current image KV into the full first-step cache by absolute position_ids.
         SP: returns cached prompt only shaped [bs, cached_prompt_len, ...] (used as joint_text).
         """
         cached_key, cached_value = self.image_kv_cache_map
         _, q_len, num_kv_heads, head_dim = key.shape
+        if cached_key.dim() == 4:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: Could you check whether this is an over-modification? In theory, the `position_ids` of subsequent steps are just suffixes of the first step. So the original logic directly cached the non-image part and reused it by concatenation.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py [hunk 2]
```diff
@@ -954,15 +958,27 @@ def _reuse_prompt_kv(
...
+            assert shard_image_size is None
+            assert position_ids is not None
+            assert position_ids.shape == (bs, q_len)
```
Comment (hsliuustc0106) [inline vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: This concat-based non-SP path (`cached_prompt_len + 1 == seq_len - q_len` ...) is now dead code. When `sp_size <= 1`, the new 4D branch at the top of the method returns early. When `sp_size > 1`, the `shard_image_size is not None` branch handles it. Consider removing this block in a follow-up to keep the method clean.

**Review by hsliuustc0106 — COMMENTED** [review]
## Review Summary

**Validated:**
- All gates pass (DCO, pre-commit, build, mergeable)
- Visual output comparison provided in PR description showing alignment with native implementation
- The three flow-level fixes (prompt whitespace, input_ids=None after first step, KV scatter via position_ids) are logically consistent with the stated goal of matching the native denoising flow
- The `input_ids=None` propagation through `prepare_inputs_for_generation` and `forward_call` is handled correctly with appropriate None-guards
- Pinning TORCH_SDPA in the DiT deploy config is reasonable for precision validation

**Non-blocking observations:**
1. The old non-SP concat path in `_reuse_prompt_kv` (lines 974-993 of the original, the `[cached_prompt, new_image, zero_eoi]` assembly) is now unreachable for `sp_size <= 1` because the new 4D branch returns first, and for `sp_size > 1` the SP branch handles it. This dead code should be cleaned up in a follow-up to avoid confusion.
2. The per-batch-element `index_copy_` loop in the new scatter path is fine for typical bs=1-2 but would benefit from a vectorized scatter if larger batches become common.

**Verdict:** No blocking issues found. The changes are well-scoped, the evidence (visual output comparison) supports the correctness claim, and the diff is consistent with the stated alignment goals.

---

*Reviewed by Claude Code with glm-5.1*

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py [hunk 3]
```diff
@@ -1219,9 +1219,14 @@ def forward_call(
         )
         custom_pos_emb = self.get_pos_emb(custom_pos_emb, position_ids)
 
-        inputs_embeds = self.model.embed_tokens(input_ids)
-
-        bsz, seq_len, n_embd = inputs_embeds.shape
+        if input_ids is not None:
```
Comment (Bounty-hunter) [inline vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1237]: Can this part be understood as the embeddings of subsequent steps being torch.cat([timestep_emb, image_emb], dim=1)? If so, we would not need input token IDs for later steps.

However, one thing I do not fully understand is that under the original logic, inputs_embeds would also be overwritten, which should theoretically be equivalent. Why would this affect accuracy?

File: vllm_omni/deploy/hunyuan_image3_dit.yaml [hunk 4]
```diff
@@ -15,6 +15,9 @@ stages:
     vae_use_tiling: false
     cache_backend:
     cache_config:
+    diffusion_attention_config:
+      default:
+        backend: TORCH_SDPA
```
Comment (Bounty-hunter) [inline vllm_omni/deploy/hunyuan_image3_dit.yaml]: I don’t think we need to set it to SDPA by default. Instead, we can let the system determine it automatically, and set SDPA at runtime through the `DIFFUSION_ATTENTION_BACKEND` parameter.

**Comment by zengchuang-hw** [comment]
Local tests are recommended to be added in the test plan.

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 1 above)
Comment (Bounty-hunter) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: After further analysis, I found that for instruct, the first-step token IDs are xxxx {timestep image} xxx, where {timestep image} is not a suffix. Perhaps we could use gen_timestep_scatter_index (officially called timesteps_index) to determine the cache size, so we can avoid caching the entire image portion together.

By the way, we also need to consider the SP scenario.

File: vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py (hunk 3 above)
Comment (bjf-frz) [reply vllm_omni/diffusion/models/hunyuan_image3/pipeline_hunyuan_image3.py:1237]: > Can this part be understood as the embeddings of subsequent steps being torch.cat([timestep_emb, image_emb], dim=1)? If so, we would not need input token IDs for later steps.
> 
> However, one thing I do not fully understand is that under the original logic, inputs_embeds would also be overwritten, which should theoretically be equivalent. Why would this affect accuracy?

Yes, this part should be theoretically equivalent because `inputs_embeds` is overwritten later. This change is not the main accuracy fix; the main accuracy issue comes from the KV cache offset mismatch. Setting `input_ids` to `None` is mainly to avoid redundant gather/embed computation in later steps and to align the flow with the native implementation.

File: vllm_omni/deploy/hunyuan_image3_dit.yaml (hunk 4 above)
Comment (bjf-frz) [reply vllm_omni/deploy/hunyuan_image3_dit.yaml]: removed

File: vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py (hunk 1 above)
Comment (bjf-frz) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: > After further analysis, I found that for instruct, the first-step token IDs are xxxx {timestep image} xxx, where {timestep image} is not a suffix. Perhaps we could use gen_timestep_scatter_index (officially called timesteps_index) to determine the cache size, so we can avoid caching the entire image portion together.
> 
> By the way, we also need to consider the SP scenario.

It used `gen_timestep_scatter_index` as the real start index of timestep and avoid setting the magic number of length of suffix

Comment (bjf-frz) [reply vllm_omni/diffusion/models/hunyuan_image3/hunyuan_image3_transformer.py]: also, it won't copy the full kvc , but only saves the prefix cache.

**Comment by Bounty-hunter** [comment]
It is a very clean and effective change. Could you also add the following tests?
（1）`pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py`
（2）`pytest -s -v tests/diffusion/models/hunyuan_image3/test_image_kv_cache_manager.py` (maybe need to change)
（3）Run the DIT only with `tp`  `sp`

**Comment by bjf-frz** [comment]
> It is a very clean and effective change. Could you also add the following tests? （1）`pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py` （2）`pytest -s -v tests/diffusion/models/hunyuan_image3/test_image_kv_cache_manager.py` (maybe need to change) （3）Run the DIT only with `tp` `sp`

Test passed.

**Comment by bjf-frz** [comment]
> It is a very clean and effective change. Could you also add the following tests? （1）`pytest -s -v tests/e2e/accuracy/test_hunyuan_image3.py` （2）`pytest -s -v tests/diffusion/models/hunyuan_image3/test_image_kv_cache_manager.py` (maybe need to change) （3）Run the DIT only with `tp` `sp`

results:

```
+-----------------------------+---------+------------------+
| Metric                      |   Value |   L20x Reference |
+=============================+=========+==================+
| COT similarity to reference |  0.9751 |           0.9644 |
+-----------------------------+---------+------------------+
| COT prefix match            | 45      |          29      |
+-----------------------------+---------+------------------+
| Image-Image similarity      | 97.2919 |          94.5538 |
+-----------------------------+---------+------------------+
| SSIM                        |  0.2838 |           0.242  |
+-----------------------------+---------+------------------+
| PSNR (dB)                   | 14.16   |          14.1    |
+-----------------------------+---------+------------------+
PASSED
```

File: tests/e2e/accuracy/test_hunyuan_image3.py [hunk 5]
```diff
@@ -78,7 +78,7 @@ def _default_ar_dit_devices() -> tuple[str, str]:
     "cot_semantic_sim": 0.9,  # Full CoT semantic similarity
     # Image comparison
     "clip_score": 85,  # CLIP image semantic similarity
-    "ssim": 0.20,  # Structural similarity
+    "ssim": 0.26,  # Structural similarity
```
Comment (Bounty-hunter) [inline tests/e2e/accuracy/test_hunyuan_image3.py:81]: I test in L20，SSIM indeed imporve, I think we can also change clip_score to 90，and PSNR to 12.5 with 10% tolerance.

```
[ONLINE] +-----------------------------+---------+------------------+
| Metric                      |   Value |   L20x Reference |
+=============================+=========+==================+
| COT similarity to reference |  0.9717 |           0.9644 |
+-----------------------------+---------+------------------+
| COT prefix match            | 29      |          29      |
+-----------------------------+---------+------------------+
| Image-Image similarity      | 95.6356 |          94.5538 |
+-----------------------------+---------+------------------+
| SSIM                        |  0.2864 |           0.242  |
+-----------------------------+---------+------------------+
| PSNR (dB)                   | 13.95   |          14.1    |
+-----------------------------+---------+------------------+
```

**Review by Bounty-hunter — APPROVED** [review]
LGTM

**Comment by bjf-frz** [comment]
@hsliuustc0106 @Gaohan123 PTAL, thx.
